certora/specs/sharesAssetsConversion.spec

```diff
@@ -0,0 +1,257 @@
+/* NOTE: The precision depends on the ratio RAY/rate.
+ */
+rule sharesConversionPreserved(uint256 shares) {
+	mathint mathshares = to_mathint(shares);
+	uint256 amount = convertToAssets(shares);
+	mathint converted = to_mathint(convertToShares(amount));
+
+	// That converted <= mathshare was proved in sharesConversionRoundedDown.
+	assert mathshares - converted <= 1 + RAY() / rate(), "Too few converted shares";
+}
+
+
+/* Joining and splitting accounts provides limited advantage.
+ * This rule verifies that joining accounts (by combining shares), and splitting accounts
+ * (by splitting shares between accounts) provides limited advantage when converting to
+ * asset amounts.
+ */
+rule accountsJoiningSplittingIsLimited(uint256 shares1, uint256 shares2) {
+    uint256 amount1 = convertToAssets(shares1);
+    uint256 amount2 = convertToAssets(shares2);
+    uint256 jointShares = shares1 + shares2;
+    require jointShares >= shares1 + shares2;  // Prevent overflow
+    uint256 jointAmount = convertToAssets(jointShares);
+
+    assert jointAmount >= amount1 + amount2, "Found advantage in combining accounts";
+    assert jointAmount < amount1 + amount2 + 2, "Found advantage in splitting accounts";
+    // The following assertion fails (as expected):
+    // assert jointAmount < amount1 + amount2 + 1, "Found advantage in splitting accounts";
+}
+
+// Similar rule as above for shares
+rule convertSumOfAssetsPreserved(uint256 assets1, uint256 assets2) {
+    uint256 shares1 = convertToShares(assets1);
+    uint256 shares2 = convertToShares(assets2);
+    uint256 sumAssets = assets1 + assets2;
+	require sumAssets >= assets1 + assets2; // Prevent overflow
+    uint256 jointShares = convertToShares(sumAssets);
+
+	assert jointShares >= shares1 + shares2, "Convert sum of assets bigger than parts";
+	assert jointShares < shares1 + shares2 + 2, "Convert sum of assets far smaller than parts";
+}
+
+// Ensure that maxWithdraw and maxRedeem are in line with the conversion functions.
+rule maxWithdrawRedeemCompliance(address owner) {
+    uint256 shares = balanceOf(owner);
+    uint256 amountConverted = convertToAssets(shares);
+
+    assert maxWithdraw(owner) <= amountConverted, "Can withdraw more than converted amount";
+    assert maxRedeem(owner) <= shares, "Can redeem more than available shares)";
+}
+
+
+// Ensure that previewWithdraw and previewRedeem are in line with the conversion functions.
+rule previewWithdrawRedeemCompliance(uint256 value) {
+    env e;
+    uint256 assets = convertToAssets(value);
+    uint256 shares = convertToShares(value);
+
+    assert previewWithdraw(e, value) >= shares, "Preview withdraw takes less shares than converted";
+    assert previewRedeem(e, value) <= assets, "Preview redeem yields more assets than converted";
+
+	// The following rule protects the client.
+    assert previewWithdraw(e, value) <= shares + 1, "Preview withdraw costs too many shares";
```

Comment:
Please move this line up. place it after the other side of the `previewWithdraw` check.